feat: UtilityProcess API (#34980)
electron/electron: docs/api/utility-process.md

> # UtilityProcess

Reviewer: When the API is fully finalized, we should also update our explainer docs (e.g. Process Model, Tutorial series) to reflect the existence of this new process type.

`UtilityProcess` creates a child process with
Node.js and Message ports enabled. It provides the equivalent of [`child_process.fork`][] API from Node.js
but instead uses [Services API][] from Chromium to launch the child process.

`UtilityProcess` is an [EventEmitter][event-emitter].

## Class: UtilityProcess

> Child process with Node.js integration spawned by Chromium.

Process: [Main](../glossary.md#main-process)<br />

> ### `new UtilityProcess(modulePath[, args][, options])`

Reviewer 2: is `new` the right way to express this? Should this be `utilityProcess.spawn()` or `process.spawnUtility()` or something else? I find it a little awkward that `new` has the side effect of spawning a process; that feels like it should be a function call. Somewhat weak opinion, though, I'm willing to be convinced otherwise.

Author: I wanted the call to represent that users have to work with an instance of the `UtilityProcess` module for every process, mostly inspired from `BrowserWindow`. But I could also wrap this in a function call which would internally just perform the same `new` call, I think an example of this would be `net.request`. Also, the module itself don't have the necessity for any static methods, all options that could modify the process behavior have to be done before the process is launched and any interaction after that will be with the process instance, so I went with the `new` interface.

Reviewer 2: Since this API is inspired by `child_process.fork()` (which does return a `ChildProcess` object), I think it makes sense for this to be a regular function rather than a constructor. I suppose our `EventEmitter` machinery requires `gin::Constructible` anyway, but that's an implementation detail as you illustrate.

I put in a weak vote for `utilityProcess.spawn()`, maybe @MarshallOfSound could weigh in and break the tie?

Reviewer 3: Whether to use constructor or factory function depends on how the APIs might be extended in future.

While this API is only about executing Node.js scripts, I think it is possible that users may want APIs to execute shell commands directly, i.e. `utilityProcess.execFile`, `utilityProcess.exec` and `utilityProcess.spwan`. The PR body also mentioned possible deno support, so maybe an `utilityProcess.forkDeno` API in future.

(I prefer `utilityProcess.fork` over `utilityProcess.spawn` for this API, because `fork` execute Node.js script, while `spawn` executes shell command).

Author: We have a tie breaker, I will rewire this to be `utilityProcess.fork`

Author: Refs https://github.com/electron/electron/pull/34980/commits/6616a24453464428b75710ac8116e0e5cc5152f8

> * `modulePath` string - Absolute path to the script that should run as entrypoint in the child process.

Reviewer 2: If we're going to require that this be a path inside the app bundle (which I think is a somewhat questionable decision, but see below), then we should document that here too.

Author: I went with the same reasoning we used for dropping support to `--require` in the `NODE_OPTIONS` to avoid injecting untrusted scripts into the process. Given the entrypoint can only be a Javascript file, wouldn't it be better if it came from within the app bundle ? I couldn't think of an example usage that would require executing scripts from outside the application given this does not support network fetch of the entry scripts. I exposed this as a strict option but we can always relax the requirement if there is usage, didn't want to start with a relaxed api and then make it strict if any security issues were discovered in real world.

Reviewer 2: Hm... what is the threat model here? What attack vector are you hypothesizing that this protects against? It's a bit unclear to me.

Author: Will be removing this check, refs https://github.com/electron/electron/pull/34980#discussion_r983884312

* `args` string[] (optional) - List of string arguments that will be available as `process.argv`
  in the child process.
* `options` Object (optional)

>   * `env` Object (optional) - Environment key-value pairs. Default is `process.env`.

Reviewer 2: are we missing `cwd`? is that not an option that is available to us? What about the other options in `child_process.fork`?

- `detached` - probably this isn't relevant for this API
- `execPath` - same
- `gid` / `uid`?
- `signal` / `killSignal`?
- `silent`?
- `windowsVerbatimArguments`?
- `timeout`?

Reviewer 2: Let's add at least `cwd`, the other non-patch-requiring things can be implemented by a user app I think.

Author: Refs https://github.com/electron/electron/pull/34980/commits/40a545b5da4496080527f11ec348eea7ce703197

  * `allowLoadingUnsignedLibraries` boolean (optional) _macOS_ - With this flag, the utility process will be
    launched via the `Electron Helper (Plugin).app` helper executable on macOS, which can be
    codesigned with `com.apple.security.cs.disable-library-validation` and
    `com.apple.security.cs.allow-unsigned-executable-memory` entitlements. This will allow the utility process
    to load unsigned libraries. Unless you specifically need this capability, it is best to leave this disabled.
    Default is `false`.

### Instance Methods

#### `child.postMessage(message, [transfer])`

* `message` any
* `transfer` MessagePortMain[] (optional)

Send a message to the child process, optionally transferring ownership of
zero or more [`MessagePortMain`][] objects.

For example:

```js
// Main process
const { port1, port2 } = new MessageChannelMain()
const child = new UtilityProcess(path.join(__dirname, 'test.js'))
child.postMessage({ message: 'hello' }, [port1])

// Child process
process.parentPort.once('message', (e) => {
  const [port] = e.ports
  // ...
})
```

#### `child.kill()`

Returns `boolean`

Terminates the process gracefully. On POSIX, it uses SIGTERM

> but will ensure to reap the process on exit. This function returns

Reviewer 2: ```suggestion
but will ensure the process is reaped on exit. This function returns
```

Author: https://github.com/electron/electron/pull/34980/commits/328ce01dadbf624898a94f448d805ed24adcb540

> true if kill succeeds, and false otherwise.

Reviewer 2: ```suggestion
true if the kill is successful, and false otherwise.
```

Author: https://github.com/electron/electron/pull/34980/commits/328ce01dadbf624898a94f448d805ed24adcb540

### Instance Properties

#### `child.pid`

A `Integer | undefined` representing the process identifier (PID) of the child process.

> If the child process fails to spawn due to errors, then the value is `undefined`.

Reviewer 2: is this still valid after the process quits?

Author: Nope, it will return undefined. I can make it explicit in the docs.

> If the child process fails to spawn due to errors, then the value is `undefined`. When

Reviewer 3: The `child_process` module would also emit `error` event when execution fails, we might need to support the same event to give users a way to handle errors.

Author: Yup I was also thinking of adding the event, it is possible to obtain this data https://source.chromium.org/chromium/chromium/src/+/main:content/browser/utility_process_host.cc;l=382

Author: I will add this event in a follow up PR, as part of hooking into launch failure notification I want to remove the use of `UtilityProcessHost` with a simpler `ElectronChildProcessHost`. Doing that in this PR will make it more difficult for review.

#### `child.stdout`

> A `NodeJS.ReadableStream | null | undefined` that represents the child process's stdout.

Reviewer 2: is this possibly a bit of a footgun? i think the default here is `pipe`, and if the parent process does not properly read the pipe then the child process will stall when the stdout or stderr buffer fills up, which can happen due to logs in chromium or the child process itself.

Author: I feel like Node.js must have the same issue given their default is also `pipe` https://nodejs.org/dist/latest-v16.x/docs/api/child_process.html#optionsstdio. I flipped the default to `inherit` https://github.com/electron/electron/pull/34980/commits/6430fc8a68285f9499df08583308c87dd5b0ff0e

If the child was spawned with options.stdio[1] set to anything other than 'pipe', then this will be `null`.
The property will be `undefined` if the child process could not be successfully spawned.

```js
// Main process
const { port1, port2 } = new MessageChannelMain()
const child = new UtilityProcess(path.join(__dirname, 'test.js'))
child.stdout.on('data', (data) => {
  console.log(`Received chunk ${data}`)
})
```

#### `child.stderr`

A `NodeJS.ReadableStream | null | undefined` that represents the child process's stderr.
If the child was spawned with options.stdio[2] set to anything other than 'pipe', then this will be `null`.
The property will be `undefined` if the child process could not be successfully spawned.

### Instance Events

#### Event: 'spawn'

Emitted once the child process has spawned successfully.

#### Event: 'exit'

Returns:

> * `event` Event

Reviewer 2: I hate that we have this `event` object everywhere. Can we delete it here? There's reason to keep it in some places for consistency (e.g. WebContents events all have `Event` as the first argument) but this API is trying to be consistent with the Node.js API, which doesn't have this pointless argument, so let's take the win and match Node here.

Author: Yup makes sense.

Author: https://github.com/electron/electron/pull/34980/commits/328ce01dadbf624898a94f448d805ed24adcb540

> * `code` number
> 
> Emitted after the child process ends. `code` contains the exit code for
> the process obtained from waitpid on posix, or GetExitCodeProcess on windows.

Reviewer 4: ```suggestion
* `code` number - Contains the exit code for
the process obtained from waitpid on POSIX, or GetExitCodeProcess on Windows.

Emitted after the child process ends.
```

#### Event: 'message'

Returns:

> * `event` Event

Reviewer 2: same here.

Author: https://github.com/electron/electron/pull/34980/commits/328ce01dadbf624898a94f448d805ed24adcb540

# UtilityProcess

`UtilityProcess` is the equivalent of [`child_process.fork`][] API from Node.js
but instead uses [Services API][] from Chromium to create child processes with

> Node.js and Message ports enabled.

Reviewer 5: Here I would maybe explain when to use utility processes and when to use node.js forked processes. I.e. advantages and disadvantages (if any). 

For VSCode the main intent why we want this API is to leverage direct renderer<->process communication via message ports.

`UtilityProcess` is an [EventEmitter][event-emitter].

## Class: UtilityProcess

> Child process with Node.js integration spawned by Chromium.

Process: [Main](../glossary.md#main-process)<br />

### `new UtilityProcess(modulePath[, args][, options])`

> * `modulePath` string - Absolute path to the script that should run as entrypoint in the child process.

Reviewer 2: This should probably match https://nodejs.org/api/worker_threads.html#new-workerfilename-options for consistency!

Author: The api is modeled after https://nodejs.org/api/child_process.html#child_processforkmodulepath-args-options since it is closer to it in behavior.

Reviewer 2: ah, a bit disappointing that Node diverges in those two APIs.

* `args` string[] (optional) - List of string arguments that will be available as `process.argv`
  in the child process.
* `options` Object (optional)
  * `env` Object - Environment key-value pairs. Default is `process.env`.
  * `execArgv` string[] - List of string arguments passed to the executable. Default is `process.execArgv`.
  * `serviceName` string - Name of the process that will appear in `name` property of
    [`child-process-gone` event of `app`](app.md#event-child-process-gone).
    Default is `node.mojom.NodeService`.
  * `disableLibraryValidation` boolean (optional) _macOS_ - Allows launching the process using

>     Helper (Plugin).app which can code signed with entitlement that disables library validation.

Author: By default generic child processes on macOS will be spawned with `Helper.app` executable, when applications have an extension architecture and the extensions can load native libraries (ex: node native modules), these libraries will not be codesigned by the application, the hardened runtime will prevent these unsigned libraries to be loaded into the process. Application has to codesign the `Helper.app` with the entitlements [com.apple.security.cs.disable-library-validation](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_disable-library-validation), [com.apple.security.cs.allow-unsigned-executable-memory](https://developer.apple.com/documentation/bundleresources/entitlements/com_apple_security_cs_allow-unsigned-executable-memory) to support this use case.

But this causes the Application main process to respect `DYLD_INSERT_LIBRARIES` environment variable which serves as untrusted code injection vector https://knight.sc/malware/2019/03/15/code-injection-on-macos.html. Chrome also has [use case](https://source.chromium.org/chromium/chromium/src/+/main:content/browser/video_capture_service.cc;l=117-122) to load unsigned libraries in certain cases and they have reserved `Helper (Plugin).app` for this specific use case https://source.chromium.org/chromium/chromium/src/+/main:content/public/common/child_process_host.h;l=116-123.

I decided to take advantage of this support and with the above flag applications can force the utility process to be launched using the `Helper (Plugin).app` and can sign only this executable with the above privileged entitlements (ex: https://source.chromium.org/chromium/chromium/src/+/main:chrome/app/helper-plugin-entitlements.plist) leaving the main entry process unaffected. I was not sure if this would be majority use case hence didn't keep this as default. Also, I am not sure how to best explain this flag in the docs with short description, open to better documentation.

>     Helper (Plugin).app which can code signed with entitlement that disables library validation.

Reviewer 2: Perhaps a better name would be `allowLoadingUnsignedLibraries`?

And document as:

> With this flag, the utility process will be launched via the `Electron Helper (Plugin).app` helper process on macOS, which by default is codesigned with the `com.apple.security.cs.disable-library-validation` and `com.apple.security.cs.allow-unsigned-executable-memory` entitlements. This will allow the utility process to load unsigned libraries. Unless you specifically need this capability, it's best to leave this disabled.

    Default is `false`.

### Instance Methods

> #### `child.postMessage(channel, message, [transfer])`

Reviewer 2: Thoughts about using `parentPort` instead, a la https://nodejs.org/api/worker_threads.html#workerparentport?

* `channel` string
* `message` any
* `transfer` MessagePortMain[] (optional)

Send a message to the child process, optionally transferring ownership of
zero or more [`MessagePortMain`][] objects.

For example:

```js
// Main process
const { port1, port2 } = new MessageChannelMain()
const child = new UtilityProcess(path.join(__dirname, 'test.js'))
child.postMessage('port', { message: 'hello' }, [port1])

// Child process

> process.on('port', (e, msg) => {

Reviewer 2: it feels a bit odd to emit an arbitrary event on `process`? It conflicts with all the existing events: https://nodejs.org/api/process.html#process-events and with any new ones that Node.js adds.

Author: Will replace with `parentPort`

  const [port] = e.ports
  // ...
})
```

#### `child.kill([signal])`

* `signal` (string | Integer) (optional)

Returns `boolean`

Sends a signal to the child process. If no argument is given,
the process will be sent the 'SIGTERM' signal. This function returns true
if kill succeeds, and false otherwise.

### Instance Properties

#### `child.pid`

A `Integer` representing the process identifier (PID) of the child process.
If the child process fails to spawn due to errors, then the value is `undefined`.

### Instance Events

#### Event: 'spawn'

Emitted once the child process has spawned successfully.

#### Event: 'exit'

Emitted when the child process has a normal termination.

> For other abnormal exit cases listen to the [`child-process-gone` event of `app`](app.md#event-child-process-gone).

Reviewer 2: it might be a nice convenience to re-emit that event on this object?

# utilityProcess

`utilityProcess` creates a child process with
Node.js and Message ports enabled. It provides the equivalent of [`child_process.fork`][] API from Node.js
but instead uses [Services API][] from Chromium to launch the child process.

Process: [Main](../glossary.md#main-process)<br />

## Methods

### `utilityProcess.fork(modulePath[, args][, options])`

* `modulePath` string - Path to the script that should run as entrypoint in the child process.
* `args` string[] (optional) - List of string arguments that will be available as `process.argv`
  in the child process.
* `options` Object (optional)
  * `env` Object (optional) - Environment key-value pairs. Default is `process.env`.

>   * `execArgv` string[] (optional) - List of string arguments passed to the executable.

Reviewer 2: Node.js defaults this to `process.execArgv`. What do we default it to?

Author: I don't think we can default to `process.execArgv`, there are two issues

1) We don't seem to populate `process.execArgv` in the main process, always returns `[]`
2) We only allow a limited set of CLI options to be propagated to Node for the main proces https://github.com/electron/electron/blob/7493062555b9eb4e6cc47d628a4831c121192f1f/shell/common/node_bindings.cc#L211-L215 so even if we fix 1) we cannot expand 2)

Can we leave the default for this option to be `[]` ?

Reviewer 2: I'm fine for the default to be whatever, but we should document what it is, & what (kind of) options can be passed here.

  * `cwd` string (optional) - Current working directory of the child process.

>   * `stdio` (string[] | string) (optional) - Child's stdout and stderr configuration. Default is `inherit`.

Reviewer 2: Node.js supports [a lot more stuff](https://nodejs.org/api/child_process.html#optionsstdio) here. Is it intentional that we support a much smaller subset of the API? If so, we should probably make that clear in the docs.

Author: Since we currently only allow configuring stdout and stderr, I choose the current subset of values since it is the most common usage for these fds. It is possible to support Stream and raw integer values, also Overlapped for windows but I feel it could be a follow-up improvement when adding support to pass arbitrary number of fds to the child process. `ipc` is omitted intentionally since there is always a channel for parent<->child communication using the postMessage API.

Author: Refs https://github.com/electron/electron/pull/34980/commits/a82d3223ede85df14b97780ef2bc1cce21e3aa39

  * `env` Object (optional) - Environment key-value pairs. Default is `process.env`.
  * `execArgv` string[] (optional) - List of string arguments passed to the executable.
  * `cwd` string (optional) - Current working directory of the child process.
  * `stdio` (string[] | string) (optional) - Allows configuring the mode for stdout and stderr
    of the child process. Default is `inherit`.
    String value can be one of `pipe`, `ignore`, `inherit`, for more details on these values you can refer to

>     [stdio][] documentation from Node.js. Currently this option only supports a subset of values listed in
>     [stdio][] documentation, support for other values will be extended when this API allows passing additional FDs
>     to the child process. Also this option does not support configuring stdin and is always set to `ignore`.

Reviewer 2: ```suggestion
    [stdio][] documentation from Node.js. Currently this option only supports configuring `stdout` and `stderr` to either `pipe`, `inherit` or `ignore`. Configuring `stdin` is not supported; `stdin` will always be ignored. 
```

Author: Refs https://github.com/electron/electron/pull/34980/commits/f6b160abb490742736220374ef703bc6015a102b
